src/stream/encoder.rs

```diff
@@ -183,34 +195,57 @@ impl<W: Write> Encoder<W> {
     ///
     /// This returns the inner writer in case you need it.
     ///
+    /// To get back `self` in case an error happened, use `try_finish`.
+    ///
     /// **Note**: If you don't want (or can't) call `finish()` manually after writing your data,
     /// consider using `auto_finish()` to get an `AutoFinishEncoder`.
     pub fn finish(mut self) -> io::Result<W> {
+        self.do_finish()?;
+        // Return the writer, because why not
+        Ok(self.writer)
+    }
 
-        // First, closes the stream.
-
-        let mut buffer = zstd_sys::ZSTD_outBuffer {
-            dst: self.buffer.as_mut_ptr() as *mut c_void,
-            size: self.buffer.capacity(),
-            pos: 0,
-        };
-        let remaining = parse_code(unsafe {
-            zstd_sys::ZSTD_endStream(self.context.s,
-                               &mut buffer as *mut zstd_sys::ZSTD_outBuffer)
-        })?;
-        unsafe {
-            self.buffer.set_len(buffer.pos);
+    /// Attempts to finish the stream. You *need* to call this after writing your stuff.
+    ///
+    /// This returns the inner writer if the finish was successful, or the
+    /// object plus an error if it wasn't.
+    ///
+    /// `write` on this object will panic after `try_finish` has been called,
+    /// even if it fails.
```

Comment:
I wonder if it'd be a good idea to, instead of returning a half-broken `Self` in the error, return a `FinishingEncoder` or something that can *only* call `finish`?